### DISCUSSION TITLE
Unable to push changes to dokku app

### DISCUSSION THREAD

**Discussion by corpschandu** [dokku/dokku#5244] [Q&A] [answered]
Hello all, 

We have a dokku app setup on our server and unfortunately we are unable to push the changes into the app and the changes are not getting reflected on the end url. Below is the summary

$ git push dokku@$APPLICATION_SERVER_STAGE_TEST:flec_pim_test HEAD:refs/heads/master --force
remote: -----> Cleaning up...        
remote: -----> Building flec_pim_test from herokuish...        
remote: -----> Adding BUILD_ENV to build environment...        
remote:        -----> Warning: Multiple default buildpacks reported the ability to handle this app. The first buildpack in the list below will be used.        
remote:               Detected buildpacks: multi nodejs        
remote:        -----> Multipack app detected        
remote: =====> Downloading Buildpack: https://github.com/dennismende/heroku-buildpack-ember-cli-yarn.git        
remote: =====> Detected Framework: Ember CLI        
remote:        Build Directory = /tmp/build        
remote:        Cache Directory = /cache        
remote: -----> Exporting config vars to environment        
remote: -----> Building environment: production        
To dokku@10.151.3.207:flec_pim_test
 ! [remote rejected] HEAD -> master (pre-receive hook declined)
error: failed to push some refs to 'dokku@10.151.3.207:flec_pim_test'
Running after_script
00:01
Running after script...
$ rm ~/.ssh/id_rsa
Cleaning up file based variables
00:01
ERROR: Job failed: exit code 1

Any suggestions or help on this is highly appreciated.

**Comment by josegonzalez** [comment]
Unless there is some hidden SIGKILL message you haven't posted, it seems like the buildpack itself is having issues. Have you tried filing a ticket upstream to see if thats the case? Did this app build before/were there any changes since the last successful build?

Does this app build successfully on Heroku?

**Reply by corpschandu** [reply]
- nvm install $NODE_VERSION
    **#- curl --insecure -o -L https://yarnpkg.com/install.sh | bash -s -- --version "$YARN_VERSION"** 
    **- npm install -g yarn@"$YARN_VERSION"**
    - export PATH=$HOME/.yarn/bin:$HOME/.config/yarn/global/node_modules/.bin:$PATH
    - cp config/environment.development.js config/environment.js
    - cp config/.ember-cli-default ./.ember-cli
    - hostname -f
    - yarn add bower
    - yarn install
    - node_modules/.bin/bower update --allow-root

Previously it was from cURL the installation used to happen as there was an issue with SSL we changed the process to install YARN from NPM .. that is the only change that we did

**Reply by josegonzalez** [reply]
So is it the bower step that fails?

**Reply by corpschandu** [reply]
Running with gitlab-runner 13.8.0 (775dd39d)
  on node_chrome_tue0[2](https://gitlab.footlocker.dmc-footlocker.noris.de/ecommerce-development/flec-service-pim-frontend/-/jobs/45633#L2)vt vzrafhxF
Preparing the "docker" executor
00:02
Using Docker executor with image gitlab.footlocker.dmc-footlocker.noris.de:5000/devops/ci/node_chrome ...
Authenticating with credentials from /root/.docker/config.json
Pulling docker image gitlab.footlocker.dmc-footlocker.noris.de:5000/devops/ci/node_chrome ...
Using docker image sha256:e7da60bf048ff891a14d[3](https://gitlab.footlocker.dmc-footlocker.noris.de/ecommerce-development/flec-service-pim-frontend/-/jobs/45633#L3)58161a5889[4](https://gitlab.footlocker.dmc-footlocker.noris.de/ecommerce-development/flec-service-pim-frontend/-/jobs/45633#L4)094c2247e18d0a0[5](https://gitlab.footlocker.dmc-footlocker.noris.de/ecommerce-development/flec-service-pim-frontend/-/jobs/45633#L5)07ea8cb14b5d554d for gitlab.footlocker.dmc-footlocker.noris.de:5000/devops/ci/node_chrome with digest gitlab.footlocker.dmc-footlocker.noris.de:5000/devops/ci/node_chrome@sha25[6](https://gitlab.footlocker.dmc-footlocker.noris.de/ecommerce-development/flec-service-pim-frontend/-/jobs/45633#L6):501200c3b410b[7](https://gitlab.footlocker.dmc-footlocker.noris.de/ecommerce-development/flec-service-pim-frontend/-/jobs/45633#L7)f6bc35e1a84cf185b24a86f[9](https://gitlab.footlocker.dmc-footlocker.noris.de/ecommerce-development/flec-service-pim-frontend/-/jobs/45633#L9)2d120685e2755147547a7ba9ac ...
Preparing environment
00:00
Running on runner-vzrafhxf-project-18-concurrent-0 via tue02vt...
Getting source from Git repository
00:03
Fetching changes...
Reinitialized existing Git repository in /builds/ecommerce-development/flec-service-pim-frontend/.git/
Checking out cd48f89a as testing-git...
Removing bower_components/
Removing config/environment.js
Removing node_modules/
Skipping Git submodules setup
Executing "step_script" stage of the job script
00:52
$ echo "$SSH_PRIVATE_KEY" | tr -d '\r' > ~/.ssh/id_rsa
$ ssh-keyscan -H $APPLICATION_SERVER_STAGE_TEST >> ~/.ssh/known_hosts
# [10](https://gitlab.footlocker.dmc-footlocker.noris.de/ecommerce-development/flec-service-pim-frontend/-/jobs/45633#L10).[15](https://gitlab.footlocker.dmc-footlocker.noris.de/ecommerce-development/flec-service-pim-frontend/-/jobs/45633#L15)1.3.207:[22](https://gitlab.footlocker.dmc-footlocker.noris.de/ecommerce-development/flec-service-pim-frontend/-/jobs/45633#L22) SSH-2.0-OpenSSH_6.7p1 Debian-5+deb8u4
# 10.151.3.207:22 SSH-2.0-OpenSSH_6.7p1 Debian-5+deb8u4
# 10.151.3.207:22 SSH-2.0-OpenSSH_6.7p1 Debian-5+deb8u4
$ chmod 600 ~/.ssh/id_rsa
$ git push dokku@$APPLICATION_SERVER_STAGE_TEST:flec_pim_uat_test HEAD:refs/heads/master --force
remote: -----> Cleaning up...        
remote: -----> Building flec_pim_uat_test from herokuish...        
remote: -----> Adding BUILD_ENV to build environment...        
remote:        -----> Warning: Multiple default buildpacks reported the ability to handle this app. The first buildpack in the list below will be used.        
remote:               Detected buildpacks: multi nodejs        
remote:        -----> Multipack app detected        
remote: =====> Downloading Buildpack: https://github.com/dennismende/heroku-buildpack-ember-cli-yarn.git        
remote: =====> Detected Framework: Ember CLI        
remote:        Build Directory = /tmp/build        
remote:        Cache Directory = /cache        
remote: -----> Exporting config vars to environment        
remote: -----> Building environment: production        
To dokku@10.151.3.207:flec_pim_uat_test
 ! [remote rejected] HEAD -> master (pre-receive hook declined)
error: failed to push some refs to 'dokku@10.151.3.207:flec_pim_uat_test'
Running after_script
00:01
Running after script...
$ rm ~/.ssh/id_rsa
Cleaning up file based variables
00:01
ERROR: Job failed: exit code 1

**Reply by corpschandu** [reply]
@josegonzalez , when building the initial dependencies the installation is getting successful 

.before-test: &before-test
  before_script:
    - export HTTP_PROXY_REQUEST_FULLURI=1
    - export HTTPS_PROXY_REQUEST_FULLURI=1
    - git config --global url.https://github.com/.insteadOf git://github.com/
    - export NODE_VERSION=$(jq -r .engines.node package.json)
    - export YARN_VERSION=$(jq -r .engines.yarn package.json)
    - echo "node version:$NODE_VERSION"
    - echo "yarn version:$YARN_VERSION"
    - nvm install $NODE_VERSION
    #- curl --insecure -o -L https://yarnpkg.com/install.sh | bash -s -- --version "$YARN_VERSION" 
    - npm install -g yarn@"$YARN_VERSION"
    - export PATH=$HOME/.yarn/bin:$HOME/.config/yarn/global/node_modules/.bin:$PATH
    - cp config/environment.development.js config/environment.js
    - cp config/.ember-cli-default ./.ember-cli
    - hostname -f
    - yarn add bower
    - yarn install
    - node_modules/.bin/bower update --allow-root
Initial dependencies are successful being completed and the job is getting succeeded


deploy::test:
  <<: *before-deploy-non-prod
  after_script:
    - rm ~/.ssh/id_rsa
  script:
    - git push dokku@$APPLICATION_SERVER_STAGE_TEST:flec_pim_test HEAD:refs/heads/master --force
  stage: deploy-test
  environment:
    name: test
    url: http://test-pim.viapim03.corp.footlocker.net/
  when: manual
  tags:
    - node

When we do this manual step of triggering the deployment into stage env we are getting this error

**Reply by josegonzalez** [reply]
I made a fork of the buildpack that turns on debugging when you set the environment variable `BUILDPACK_XTRACE=1`: https://github.com/dennismende/heroku-buildpack-ember-cli-yarn/pull/1

Can you set that env var and try my version to see if it spits out some debugging information? Please create a gist on gist.github.com of the output and paste the link here to your gist (that will make it easier to read).

**Reply by corpschandu** [reply]
@josegonzalez , link to the gist https://gist.github.com/corpschandu/a8b1148bbe0f47bd3823cecd20f284c6 
I created this file using the command **dokku trace on** as i do not have access to the repo that links to the buildpack .. 

Please have a look and let me know if that works.

**Reply by josegonzalez** [reply]
Can you switch to my buildpack fork? `https://github.com/josegonzalez/heroku-buildpack-ember-cli-yarn#patch-1` should be your buildpack url. Please also then set the env var as I mentioned and post the build output in a gist. Doing `dokku trace:on` doesn't help.

**Reply by corpschandu** [reply]
https://github.com/josegonzalez/heroku-buildpack-ember-cli-yarn/tree/patch-1 - this is where i see the change and in my gitlab.yml file i added a new var as 
.before-test: &before-test
  before_script:
    - export BUILDPACK_XTRACE=1
    - echo $BUILDPACK_XTRACE;
    - export HTTP_PROXY_REQUEST_FULLURI=1
 If this is fine i can go ahead and push the changes and run 
Please suggest

**Reply by josegonzalez** [reply]
Can you post your `.buildpacks` file?

Also, that env var needs to be set on your app, not in the CI pipeline. That means via `dokku config:set`.

**Reply by corpschandu** [reply]
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/100302224/178551535-518f9447-451d-40c4-9902-6038baa7ddba.png">

I have set the dokku config:set flec_pim_test BUILDPACK_XTRACE=1

**Reply by josegonzalez** [reply]
The contents of that file need to be:

```
https://github.com/josegonzalez/heroku-buildpack-ember-cli-yarn#patch-1
```

Please change it to that and gist the output of the proceeding deploy.

**Reply by corpschandu** [reply]
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/100302224/178553904-df381f95-0837-4756-839b-ba3a31c8f309.png">

I changed the content of the buildpack file and pushed the changes into gitlab adding the urls for your reference.. 

https://gist.github.com/corpschandu/84bfb8978795a1f9ca81e89acdbce79e - Initial dependencies
https://gist.github.com/corpschandu/742d297ec1735d62d798fb9521a0483e - after buildpack change

**Reply by josegonzalez** [reply]
Seems like this endpoint is responding with empty string or similar:

```
curl --silent --get --data-urlencode range=8.11.2 https://semver.io/node/resolve 
```

Can you run that command from your server? Maybe whatever that service is has blocked your host?

**Reply by corpschandu** [reply]
<img width="864" alt="image" src="https://user-images.githubusercontent.com/100302224/178703673-6ceafb8b-9e3c-4cd5-b90f-a5bf1c42e300.png">

This is the result i see after running the curl command ..

**Reply by corpschandu** [reply]
<img width="879" alt="image" src="https://user-images.githubusercontent.com/100302224/178712807-39bfebb2-e416-4e9d-8e25-4ea8a8cbe6e6.png">

This is the result for the Yarn command in my local machine

**Reply by corpschandu** [reply]
@josegonzalez , have you got a chance to check the above screenshots .. 

Attaching another gist file which we saw when we initially ran using curl : https://gist.github.com/corpschandu/b748493e28fa990dcccf8ef966b91d45

if the ssl failed in this situation then can it be a possibility that the compile is also failing while fetching anything from curl?
adding screenshot of ssl failute for your reference:
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/100302224/178749179-bc4f60d4-5d31-41f9-b510-c200d04694bb.png">

If curl is the problem, then is there any other way we can get the packages that are using curl to download them into the server?

**Reply by corpschandu** [reply]
Hi @josegonzalez, I created a new gist file and identified some issues, I forked the same repo into my user and made some modifications, please look into it and suggest on the same - https://github.com/corpschandu/heroku-buildpack-ember-cli-yarn https://gist.github.com/corpschandu/de9e2cc09b62f1d53b6636c719797b5f

Looking forward for your reply on it, Thanks for the support.

**Reply by josegonzalez** [reply]
That pcre file in the logs, can you run that on your host? It seems the file is accessible to me, but maybe you need to follow redirects on it?

**Reply by corpschandu** [reply]
@josegonzalez, I tried running 
curl --insecure https://github.com/PCRE2Project/pcre2/releases/download/pcre2-10.35/pcre2-10.35.tar.gz | tar xzf - -C /tmp/ 
in my local machine and i see the below 

<img width="1210" alt="image" src="https://user-images.githubusercontent.com/100302224/179196637-49434312-607f-4777-9fde-a8f09516e4ac.png">

Please suggest.

**Reply by josegonzalez** [reply]
I would try adding an allowed redirect in your fork of the file, or write the file to disk (while showing any errors) and then extract it. Seems like the pcre download is the bit that fails.

**Reply by corpschandu** [reply]
Hello @josegonzalez , the problem was with the https://ftp.pcre.org which is no longer accessible .. I modified the url to get the pcre2 package from https://ftp.exim.org after which the package got fetched into the process and I am able to do a **successful deployment** now .. I really appreciate your support on this and will get back if anything else is required and we can close this thread .. Have a good one .. Thanks

**Comment by corpschandu** [comment]
@josegonzalez 
https://github.com/dennismende/heroku-buildpack-ember-cli-yarn -- this is the buildpack that we are using and this is an old project developed by one of our former developers and not much idea .. any suggestions will be absolutely great as we are unable to make changes to the frontend application due to the pre-receive hook declined .. 

The last time there was a successful deployment 1 year back and after that there was no deployment at all.

**Comment by josegonzalez** [comment, marked as answer]
The problem here was the buildpack specified by the user had outdated urls for it's compile-time dependencies. With updated urls in place, the build succeeded.
